### ISSUE TITLE
🐛 BUG: AstroUserConfig.base does not apply to assets in public folder 

### ISSUE THREAD

**Issue by daymxn** [withastro/astro#4128]
### What version of `astro` are you using?

1.0.0-rc.3

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Windows

### Describe the Bug

The property `base` on `AstroUserConfig` does not apply to assets in the public folder. This made for some fun debugging with Github pages. I'm not sure if this is by design or not though. If it's by design, a note should probably be made in the documentation about it.

For example, given the base config of:
```
import { defineConfig } from 'astro/config';

export default defineConfig({
    site: "https://astro.github.io",
    base: "/my-project"
});
```

A file in the public dir as such `public/make-scrollable-code-focusable.js`  would compile to:
```
<script src="/make-scrollable-code-focusable.js">
```
When the expected result is as such:
```
<script src="/my-project/make-scrollable-code-focusable.js">
```

### Link to Minimal Reproducible Example

https://astro.new/docs?on=stackblitz

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by delucis** [comment]
Hi @daymxn! Would you be able to provide a link to your code reproduction so we can debug this? Looks like you're reproduction link is just the default docs starter.

**Comment by ColinEspinas** [comment]
Hello @delucis,

This is a big issue for us as we need to deploy on a custom base. 

To add a bit on what @daymxn said, it seems that any link in the template of a page is not updated by the base when building. 

As an exemple any `<a>` tag href nor `<img>` tag src is not updated either.

Here is a reproduction with comments to help you figure out what is happening and what should be happening:
https://stackblitz.com/edit/github-xf1evk?file=src/pages/index.astro

I hope this can help fix this issue. Have a great day.

**Comment by ColinEspinas** [comment]
Still in `needs-repro` any update @delucis?

Have a great day.

**Comment by matthewp** [comment]
Astro does not modify your HTML. If your script tags need a base you need to add it to the tag yourself, just directly write:

```js
<script src="/my-project/make-scrollable-code-focusable.js">
```

This should work in dev and prod, if it doesn't let me know.

**Comment by matthewp** [comment]
Going to go ahead and close this, but happy to continue chatting here.

**Comment by kndshein** [comment]
@matthewp I came across [this thread](https://github.com/withastro/roadmap/discussions/292#discussioncomment-3773908), which seemed to work really well, until I got to sub-paths, where the links immediately broke. The reason being that I use a Layout.astro component that contains all the `<head>` values, including a reference to the CSS file. So, when it was no longer one layer away from the root path, the relative link was pointing at the wrong layer.

As a result, I'm now pivoting back to the current thread. Is it really the best practice to prefix _all_ links w/ `import.meta.env.BASE_URL`, like so:
```JS
<script src={`${import.meta.env.BASE_URL}/my-project/make-scrollable-code-focusable.js`}>
```
I could abstract it out to a small util function that accepts a string, but even so, it seems like a lot of dev friction to have to do that:
```JS
<img src={prefixPath(<img_path_here>)} alt="placeholder"/>
```

If that is the case, no complaints! But I'm curious to hear if I'm missing a magical config that automatically does it for me or a silver bullet somewhere. (P.S. I tried [astro-relative-links](https://github.com/ixkaito/astro-relative-links) to no avail.)
